### PR TITLE
Support wires as variable objects

### PR DESCRIPTION
**Pull request by josh146**
**Description of the Change:**

This PR supports wires being `Variable` objects.

**Benefits:**
This allows us to support the use case where the wire number may be passed as a keyword argument.

**Possible Drawbacks:**

None, as far as I can tell.

**Related GitHub Issues:**

#150

### CODE REVIEW

File: pennylane/qnode.py [hunk 1]
```diff
@@ -285,6 +285,7 @@ def construct(self, args, **kwargs):
             raise QuantumFunctionError('QNode._current_context must not be modified outside this method.')
         # generate the program queue by executing the quantum circuit function
         try:
+            Variable.kwarg_values = keyword_values
```
Comment (co9olguy) [inline pennylane/qnode.py]: Trying to understand, what is this assignment doing?

File: pennylane/operation.py [hunk 2]
```diff
@@ -285,7 +286,7 @@ def __init__(self, *args, wires=None, do_queue=True):
             assert self.grad_recipe is None, 'Gradient recipe is only used by the A method!'
 
         # apply the operation on the given wires
-        if isinstance(wires, int):
+        if not isinstance(wires, Sequence):
             wires = [wires]
```
Comment (cgogolin) [inline pennylane/operation.py]: Good. That's an improvement.

File: tests/test_qnode.py [hunk 3]
```diff
@@ -460,6 +461,20 @@ def circuit(w, x=None):
...
+        self.logTestName()
+
+        def circuit(x, q=None):
+            qml.RX(x, [q])
+            return qml.expval.PauliZ(q)
+
+        circuit = qml.QNode(circuit, self.dev1)
+
+        c = circuit(np.pi, q=0)
+        self.assertAlmostEqual(c, -1., delta=self.tol)
```
Comment (co9olguy) [inline tests/test_qnode.py]: can we also add a test where the default of `q` is used? (would have to be a valid value, not `None`)

Comment (josh146) [reply tests/test_qnode.py]: Sure, I'll do that now

File: tests/test_qnode.py [hunk 4]
```diff
@@ -460,6 +461,20 @@ def circuit(w, x=None):
         self.assertAllAlmostEqual(c, [-1., -1.], delta=self.tol)
 
 
+    def test_keywordargs_for_wires(self):
+        "Tests that wires use keyword arguments."
```
Comment (cgogolin) [inline tests/test_qnode.py]:
```suggestion
        "Tests that wires can be passed as keyword arguments."
```

File: pennylane/qnode.py (hunk 1 above)
Comment (josh146) [reply pennylane/qnode.py]: When the `Operation` class queries for the value of a Variable, it is essentially replacing the variable identifier with the resulting value from the `Variable.kwarg_values` class attribute (shared by all Variable instances). Without this line here, the `Variable.kwarg_values` is never set, so on construction, the variable value cannot be returned (an exception is raised).

File: pennylane/qnode.py [hunk 5]
```diff
@@ -499,6 +482,23 @@ def evaluate(self, args, **kwargs):
...
+            """Make sure only existing wires are referenced."""
+            for w in op.wires:
+                if w < 0 or w >= self.num_wires:
+                    raise QuantumFunctionError("Operation {} applied to wire {}, "
+                                               "device only has {}.".format(op.name, w, self.num_wires))
+
+        # check every gate/preparation and ev measurement
+        for op in self.ops:
+            check_op(op)
+
```
Comment (josh146) [inline pennylane/qnode.py]: Since `wires` can now be `Variables`, all validation must be moved from `qnode.construct` (since keyword arguments are unknown on construction) to `qnode.evaluate` (where they will be known).

File: pennylane/operation.py [hunk 6]
```diff
@@ -341,6 +360,21 @@ def check_domain(self, p, flattened=False):
...
+        :class:`~.variable.Variable` instances are replaced by their
+        current numerical value.
+
+        Returns:
+            list[int]: wire values
+        """
+        w = [i.val if isinstance(i, Variable) else i for i in self._wires]
+        self.check_wires(w)
+        return w
+
```
Comment (josh146) [inline pennylane/operation.py]: Since wires can now be Variables, they need to be dynamically queried, instead of stored in a static attribute.

File: pennylane/operation.py [hunk 7]
```diff
@@ -285,24 +286,42 @@ def __init__(self, *args, wires=None, do_queue=True):
...
-        if isinstance(wires, int):
-            wires = [wires]
-
-        if self.num_wires != 0 and len(wires) != self.num_wires:
-            raise ValueError("{}: wrong number of wires. "
-                             "{} wires given, {} expected.".format(self.name, len(wires), self.num_wires))
+        if not isinstance(wires, Sequence):
+            self._wires = [wires]
+        else:
+            self._wires = wires
```
Comment (josh146) [inline pennylane/operation.py]: The new `self._wires` attribute stores the wires as a sequence of both `ints` and `Variables`. To dynamically determine the list of wires as integers, the new `self.wires` property (replacing the old `self.wires` attribute, but returning the same result for compatibility) must be accessed

File: pennylane/operation.py [hunk 8]
```diff
@@ -285,24 +286,42 @@ def __init__(self, *args, wires=None, do_queue=True):
...
 
-        if len(set(wires)) != len(wires):
-            raise ValueError('{}: wires must be unique, got {}.'.format(self.name, wires))
+        if all([isinstance(w, int) for w in self._wires]):
+            # If all wires are integers (i.e., not Variable), check
+            # that they are valid for the given operation
+            self.check_wires(self._wires)
```
Comment (josh146) [inline pennylane/operation.py]: We still perform wire validation on construction **if** they do not contain variables.

File: pennylane/qnode.py [hunk 9]
```diff
@@ -501,6 +482,23 @@ def evaluate(self, args, **kwargs):
...
+        m_wires = list(w for ex in self.ev for w in ex.wires)
+        if len(m_wires) != len(set(m_wires)):
+            raise QuantumFunctionError('Each wire in the quantum circuit can only be measured once.')
+
+        def check_op(op):
+            """Make sure only existing wires are referenced."""
+            for w in op.wires:
+                if w < 0 or w >= self.num_wires:
+                    raise QuantumFunctionError("Operation {} applied to wire {}, "
+                                               "device only has {}.".format(op.name, w, self.num_wires))
```
Comment (co9olguy) [inline pennylane/qnode.py]: let's update assertion text. "device only has {}" may not make sense, especially if w<0 is given.
